Added uuid4 Function

Fleshgrinder opened the pull request:
* Added freestanding `uuid4` function (see #29)
* Added JVM name `UuidUtil` for freestanding functions
* Deprecated no-args constructor in favor of `uuid4`

Fleshgrinder committed:
Added uuid4 Function

* Added freestanding `uuid4` function (see #29)
* Added JVM name `UuidUtil` for freestanding functions
* Deprecated no-args constructor in favor of `uuid4`

benasher44 commented on src/commonTest/kotlin/UuidTest.kt:
```diff
@@ -44,9 +44,8 @@ class UuidTest {
-    @Test
-    fun generates_a_UUID_with_correct_version_and_variant_bits() {
-        val uuid = Uuid()
+    @Test fun uuid4() {
```
This should be renamed to something like generates_a_uuid4 to avoid naming conflict with the new uuid4 function

Fleshgrinder replied:
A naming conflict is not possible, this function is part of the `UuidTest` class.

Fleshgrinder replied:
😝 of course, yeah.

Fleshgrinder committed:
Fix naming conflict

benasher44 commented on src/commonMain/kotlin/com/benasher44/uuid/uuid.kt:
```diff
@@ -181,3 +185,35 @@ public class Uuid(val uuid: ByteArray = genUuid()) {
+ * source of the platform.
+ *
+ * @return New version 4 [UUID][Uuid] of random data.
+ * @sample UuidTest.uuid4_generation
```
Whoops looks like the find/replace got one too many :)

Fleshgrinder replied:
The name is correct but `UuidTest` does not reside in a package and I didn't use auto-import from IntelliJ because the current EAP has huge problems…

benasher44 commented:
Looks like Jvm can't find uuid4 in commonTest. Should be able to reproduce by running the compileTestKotlinJvm task locally

Fleshgrinder committed:
Moved UuidTest into package

Also removed the unnecessary subdirectories as it is actually recommended by JetBrains.

Fleshgrinder committed:
Merge remote-tracking branch 'upstream/master' into uuid4

Fleshgrinder committed:
Updated CHANGELOG closes #41

Fleshgrinder commented:
Also updated the CHANGELOG (#41) and will provide updates to it with every PR so that you don’t have to keep track of that in the future. Would maybe be a good thing in a `CONTRIBUTING.md` to ask for. 😉

benasher44 commented:
Ah good point! Will add an issue so I remember to update it

benasher44 commented:
Now it looks like there might be a test failure. It might also be nice to fix the new deprecation warnings in tests, now that UUID constructor is deprecated

Fleshgrinder committed:
Updated Uuid() to uuid4()

Fleshgrinder commented:
I get the following if I try to run `gradle check` locally, no clue (yet).

```
gradle check

> Configure project :
Kotlin Multiplatform Projects are an experimental feature.
Target 'mingw' for platform mingw_x64 is ignored during build on this macos_x64 machine. You can build it with a mingw_x64 host.

> Task :compileKotlinLinux64 FAILED
e: Could not find "stdlib" in [/Users/richardfussenegger/Documents/Projects/github.com/Fleshgrinder/uuid, /Users/richardfussenegger/.konan/klib, /Users/richardfussenegger/.konan/kotlin-native-macos-1.3/klib/common, /Users/richardfussenegger/.konan/kotlin-native-macos-1.3/klib/platform/linux_x64].

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlinLinux64'.
> Process 'command '/Library/Java/JavaVirtualMachines/jdk1.8.0_202.jdk/Contents/Home/bin/java'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
4 actionable tasks: 1 executed, 3 up-to-date
```

benasher44 commented:
What kind of machine are you using? #45 might help once it lands. it has some machine checks to restrict windows and linux to their own hosts, respectively

benasher44 commented:
Okay #45 has landed. Try updating?

benasher44 commented:
Oh `ktlint` is failing. Try just running `./gradlew ktlintformat`

Fleshgrinder committed:
Merge remote-tracking branch 'upstream/master' into uuid4

Fleshgrinder committed:
Fixed lint issues

Fleshgrinder commented:
@benasher44 still failing, I have a MacBook Pro. It worked not so long ago to run all tests.

benasher44 commented:
I just tried myself (also a MBP) and `./gradlew check` passed for me on your branch. There are still some warnings in `CocoaUuidTest.kt` (more UUID -> Uuid), which would be nice to fix though :).

Looks like CI is back to ✅. Do you want to fix those warnings now or as follow up?

Fleshgrinder committed:
Fixed more small issues

Fleshgrinder commented on src/commonMain/kotlin/uuid.kt:
```diff
@@ -139,8 +144,8 @@ public class Uuid(val uuid: ByteArray = genUuid()) {
-                var i = range.start
-                while (i < range.endInclusive) {
+                var i = range.first
+                while (i < range.last) {
```
This is better because `first` and `last` give the unboxed types.

Fleshgrinder commented:
@benasher44 reinstalled Konan and `gradle check` works now. What do you mean with _too much find/replace_? All looks good for `gradle check` and CI agrees.

benasher44 commented on src/commonMain/kotlin/uuid.kt:
```diff
@@ -76,11 +80,11 @@ public class Uuid(val uuid: ByteArray = genUuid()) {
      *
      * @return The version number of this [Uuid].
-     * @sample UUIDTest.versions
+     * @sample com.benasher44.uuid.UuidTest.versions
```
This doc doesn't seem right (same with variants)

Fleshgrinder replied:
IntelliJ (although EAP and probably not a good benchmark right now because latest EAP has some issues) cannot resolve it without for no good reason. 😋

benasher44 commented:
Sorry that last comment was confusing. I left a better comment inline

benasher44 commented:
We can circle back once the next IntelliJ is out.
